lib/WUI/espif.cpp

```diff
@@ -0,0 +1,538 @@
+                rx_read = 0;
+            }
+
+            // Filled all pbufs in a packet (current set to next = NULL)
+            if (!rx_buff_cur) {
+                log_debug(ESPIF, "Read packet size: %d", rx_len);
+                if (netif->input(rx_buff, netif) != ERR_OK) {
+                    log_error(ESPIF, "ethernetif_input: IP input error");
+                    pbuf_free(rx_buff);
+                    state = Intron;
+                    break;
+                }
+                log_debug(ESPIF, "Input packet processed ok");
+                state = Intron;
+            }
+        } break;
+        }
+    }
+    log_debug(ESPIF, "Processed %d from UART", size);
+}
+
+/**
+ * @brief Send packet using ESPIF NIC
+ *
+ * @param netif Output NETIF handle
+ * @param p buffer (chain) to send
+ */
+static err_t low_level_output(struct netif *netif, struct pbuf *p) {
+    if (esp_operating_mode != ESPIF_RUNNING_MODE) {
+        log_error(ESPIF, "Cannot send packet, not in running mode.");
+        return ERR_IF;
+    }
+
+    const uint32_t len = p->tot_len;
+    log_debug(ESPIF, "Low level output packet size: %d", len);
+
+    xSemaphoreTake(uart_write_mutex, portMAX_DELAY);
+    espif_transmit_data(INTRON, sizeof(INTRON));
+    uint8_t msg_type = MSG_PACKET;
+    espif_transmit_data(&msg_type, 1);
+    espif_transmit_data(&len, sizeof(len));
+    while (p != NULL) {
+        if (espif_transmit_data(p->payload, p->len) != ERR_OK) {
+            log_error(ESPIF, "Low level output packet failed");
+            xSemaphoreGive(uart_write_mutex);
+            return ERR_IF;
+        }
+        p = p->next;
+    }
+
+    xSemaphoreGive(uart_write_mutex);
+    return ERR_OK;
+}
+
+static err_t reset_and_wait() {
+    // Capture this task handle to manage wakeup from input thread
+    init_task_handle = xTaskGetCurrentTaskHandle();
+
+    // Reset device to receive MAC address
+    log_debug(ESPIF, "Resetting ESP to receive MAC address");
+    hard_reset_device();
+
+    // Wait for esp ready (receiveing device info)
+    if (!xTaskNotifyWait(0, 0, NULL, pdMS_TO_TICKS(3000))) {
+        log_error(ESPIF, "Timeout waiting for ESP bootup");
+        return ERR_IF;
+    }
+    log_info(ESPIF, "ESP up and running");
+    return ERR_OK;
+}
+
+/**
+ * @brief Initalize ESPIF network interface
+ *
+ * This initializes NET interface. Can be called multiple times if ESP modes are switched.
+ *
+ * @param netif Interface to initialize
+ * @return err_t Possible error encountered during initialization
+ */
+err_t espif_init(struct netif *netif) {
+    log_info(ESPIF, "LwIP init");
+
+    espif_reconfigure_uart(NIC_UART_BAUDRATE);
+    esp_operating_mode = ESPIF_RUNNING_MODE;
+
+    // Create mutex to protect UART writes
+    uart_write_mutex = xSemaphoreCreateMutex();
+    if (!uart_write_mutex) {
+        return ERR_IF;
+    }
+
+    // Create UART reader thread
+    uart_reader_id = osThreadCreate(osThread(UartBufferThread), netif);
```

Comment:
Similar, what happens to the previous thread?

Reply:
Sorry, I forgot to update the documentation. The espif_init is not supposed to be called again. Documentation fixed + check added to avoid subsequent calls.